ecosystem/sep-0011.md

```diff
@@ -0,0 +1,201 @@
+
+### Fields
+
+A field has the following syntax:
+
+> *field* = *tag* *selector*\* \[*pseudoselector*\]
+> 
+> *selector* = `.` *tag* | `[` *integer* `]`
+> 
+> *tag* = *letter* *wordchar*\*
+> 
+> *pseudoselector* = `.` `_inner`\* `_present` | `.len`
+> 
+> *integer* = A decimal integer
+> 
+> *letter* = Any letter
+> 
+> *wordchar* = Any letter, any digit, or `_`
+
+Disregarding *pseudoselector*, each *field* names a field in an [XDR `TransactionEnvelope` data structure](https://github.com/stellar/stellar-core/blob/master/src/xdr/Stellar-transaction.x#L351). The name is generated by joining XDR field names with a period (`.`), similar to the syntax used for accessing nested fields in C++ or go representations of XDR data structures. Array elements (for both fixed- and variable-length arrays) are indexed using square brackets. As in C, arrays are 0-based. Pseudoselectors of the form `_inner_present` (or `_inner_inner_present`, etc.) are reserved for fields of type pointer to pointer. Such types do not appear in the current version of `TransactionEnvelope`, but might in the future, or txrep might be applied to other XDR data types that contain such nested pointers.
+
+As an example, the *field* `tx.timeBounds.minTime` names the `minTime` field in the `tx` field of the `TransactionEnvelope` structure, which looks like this:
+
+``` c
+struct TimeBounds {
+    uint64 minTime;
+    uint64 maxTime; // 0 here means no maxTime
+};
+struct Transaction {
+    /* ... */
+    TimeBounds* timeBounds;
+    Operation operations<100>;
+    /* ... */
+};
+struct TransactionEnvelope {
+    Transaction tx;
+    DecoratedSignature signatures<20>;
+};
+```
+
+Pointers and variable-length arrays use pseudoselectors to describe their state.
+
+Pointers use the `._present` pseudoselector with value `true` or `false` to indicate that the field is present or NULL, respectively. For example, a transaction with timebounds might be specified like this (using comments to annotate the times):
+
+    tx.timeBounds._present: true
+    tx.timeBounds.minTime: 1535756672 (Fri Aug 31 16:04:32 PDT 2018)
+    tx.timeBounds.maxTime: 1567292672 (Sat Aug 31 16:04:32 PDT 2019)
+
+A transaction without timebounds would contain this line:
+
+    tx.timeBounds._present: false
+
+For fields of type pointer-to-pointer, the pseudoselector `._present` describes whether the outermost pointer is NULL or not. If `._present` is `true`, then `._inner_present` describes whether the nested pointer is NULL or not. `_inner` can be repeated as many times as needed for arbitrarily nested pointers.
+
+Variable-length arrays use the pseudo-selector `.len` with an integer value to indicate the number of elements of the variable-length array. Only indices from 0 to len-1 are meaningful. For example, a transaction with one operation might look like this:
+
+    tx.operations.len: 1
```

Comment:
This potentially opens many issues with memory allocation we observed in XDR parsers. I can write:
```
tx.operations.len: 2147483647
```
and allocate big part of computer's memory. This is especially dangerous on servers that need to parse txrep transactions.

Not sure if it's the right place but we should state that parsers **must** check the type max length before allocating the memory. Maybe "Writing parser guidelines" section?